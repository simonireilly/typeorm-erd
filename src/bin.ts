#!/usr/bin/env node

import { ERDBuilder } from "./index";
import { resolve } from "path";

const main = async (path: string = "./examples/basic/src/data-source") => {
  const filePath = resolve(path);
  console.log("CWD Stats:", { cwd: process.cwd(), filePath });

  const moduleA = await import(filePath);
  console.info("Module", moduleA);
  const erd = new ERDBuilder("mermaid", moduleA);
  await erd.initialize();
  const erdText = await erd.render();

  console.info(erdText);
};

main();
