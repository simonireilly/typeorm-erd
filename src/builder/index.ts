import { DataSource, EntityMetadata, EntitySchema } from "typeorm";
import { ConnectionMetadataBuilder } from "typeorm/connection/ConnectionMetadataBuilder";
import { RelationType } from "typeorm/metadata/types/RelationTypes";

export type BuilderRelations = Record<
  string,
  {
    entityRelations: RelationData[];
  }
>;

interface RelationData {
  relationType: RelationType;
  propertyPath: string;
  nullable: boolean;
  isOwning: boolean;
  inverseSidePropertyPath: string;
  source: string;
  target: string;
}

const entityMetaData = async (connection: DataSource) => {
  const connectionMetadataBuilder = new ConnectionMetadataBuilder(connection);

  const { entities } = connection.options;
  const TEntities = entities as (Function | EntitySchema<any> | string)[];

  let entityMetadata;

  if (entities) {
    entityMetadata = await connectionMetadataBuilder.buildEntityMetadatas(
      TEntities
    );
  } else {
    throw Error("No entities found on connection");
  }

  return entityMetadata;
};

/**
 * Return the relation data for each table
 *
 */
const relations = (meta: EntityMetadata[]): BuilderRelations => {
  const relationData = meta.reduce<BuilderRelations>((acc, entity) => {
    const entityRelations = entity.ownRelations.map(
      ({
        relationType,
        inverseEntityMetadata,
        propertyPath,
        inverseSidePropertyPath,
        isOwning,
      }) => {
        return {
          relationType,
          propertyPath,
          isOwning,
          nullable:
            entity.columns.find((c) => c.propertyName === propertyPath)
              ?.isNullable || false,
          inverseSidePropertyPath,
          source: entity.tableName,
          target: inverseEntityMetadata.tableName,
        };
      }
    );

    return {
      ...acc,
      [entity.tableName]: {
        entityRelations,
      },
    };
  }, {});

  return relationData;
};

export const builders = {
  entityMetaData,
  relations,
};
