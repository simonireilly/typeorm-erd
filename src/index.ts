export * from "./builder";
export * from "./adapters";
