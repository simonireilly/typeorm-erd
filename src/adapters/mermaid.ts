import { DataSource, EntityMetadata } from "typeorm";
import { RelationMetadata } from "typeorm/metadata/RelationMetadata";
import { BuilderRelations, builders } from "../builder";

const RelationShips = {
  left: {
    "one-to-many": "||",
    "many-to-one": "}|",
    "one-to-one": "||",
    "many-to-many": "}|",
  },
  right: {
    "one-to-many": "|{",
    "many-to-one": "||",
    "one-to-one": "||",
    "many-to-many": "|{",
  },
};

export class Mermaid {
  private meta: EntityMetadata[];
  private relations: BuilderRelations;

  constructor(
    readonly dataSource: DataSource,
    private entityBuilder = builders.entityMetaData,
    private relationBuilder = builders.relations
  ) {
    this.dataSource = dataSource;
    this.entityBuilder = entityBuilder;
  }

  public async initialize() {
    this.meta = await this.entityBuilder(this.dataSource);
    this.relations = await this.relationBuilder(this.meta);
  }

  public render(): string {
    return `erDiagram\n  ${this.renderTables()}\n  ${this.renderRelations()}`;
  }

  public renderRelations() {
    return this.buildRelations().join("\n  ");
  }

  public buildRelations() {
    return Object.entries(this.relations).reduce((acc, [key, values]) => {
      const relations = values.entityRelations.map((rel) => {
        // Remove unnamed relations and only add relations for the explicit owner
        if (!rel.propertyPath || !rel.isOwning) return "";

        return `${rel.source} ${RelationShips["left"][rel.relationType]}--${
          RelationShips["right"][rel.relationType]
        } ${rel.target}: ${rel.propertyPath}`;
      });

      return [...acc, ...relations.filter(Boolean)];
    }, [] as string[]);
  }

  private renderTables() {
    return this.meta
      .map((entry) => {
        const columns = entry.columns.map((column) => {
          return [
            this.dataSource.driver.normalizeType(column),
            column.databaseName,
            column.isPrimary ? "PK" : column.referencedColumn ? "FK" : "",
            column?.comment && `"${column?.comment}"`,
          ]
            .filter(Boolean)
            .join(" ");
        });

        return `${entry.tableName} {\n    ${columns.join("\n    ")}\n  }`;
      })
      .join("\n  ");
  }
}
