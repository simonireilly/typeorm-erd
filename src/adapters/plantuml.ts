import { DataSource, EntityMetadata } from "typeorm";
import { BuilderRelations, builders } from "../builder";

const RelationShips = {
  left: {
    "one-to-many": "||",
    "many-to-one": "}|",
    "one-to-one": "||",
    "many-to-many": "}|",
  },
  right: {
    "one-to-many": "|{",
    "many-to-one": "||",
    "one-to-one": "||",
    "many-to-many": "|{",
  },
};

/**
 * Render a plantuml string based on the specification here:
 * https://plantuml.com/ie-diagram
 */
export class PlantUMLErd {
  private meta: EntityMetadata[];
  private relations: BuilderRelations;

  constructor(
    readonly dataSource: DataSource,
    private entityBuilder = builders.entityMetaData,
    private relationBuilder = builders.relations
  ) {
    this.dataSource = dataSource;
    this.entityBuilder = entityBuilder;
  }

  public async initialize() {
    this.meta = await this.entityBuilder(this.dataSource);
    this.relations = await this.relationBuilder(this.meta);
  }

  public render(): string {
    return `@startuml\nskinparam linetype ortho\n${this.renderTables()}\n${this.renderRelations()}\n@enduml`;
  }

  public renderRelations() {
    return this.buildRelations().join("\n");
  }

  public buildRelations() {
    return Object.entries(this.relations).reduce((acc, [key, values]) => {
      const relations = values.entityRelations.map((rel) => {
        // Remove unnamed relations and only add relations for the explicit owner
        if (!rel.propertyPath || !rel.isOwning) return "";

        return `${rel.source} ${RelationShips["left"][rel.relationType]}--${
          RelationShips["right"][rel.relationType]
        } ${rel.target}`;
      });

      return [...acc, ...relations.filter(Boolean)];
    }, [] as string[]);
  }

  private renderTables() {
    return this.meta
      .map((entry) => {
        const columns = entry.columns.map((column) => {
          const { databaseName, isPrimary, referencedColumn, comment, length } =
            column;

          const normalizedColumnName = this.dataSource.driver
            .normalizeType(column)
            .toUpperCase();
          const columnName = length
            ? `${normalizedColumnName}(${length})`
            : normalizedColumnName;

          return [
            databaseName,
            columnName,
            isPrimary ? "<<PK>>" : referencedColumn ? "<<FK>>" : "",
            comment && `//"${comment}"//`,
          ]
            .filter(Boolean)
            .join(" : ");
        });

        return `entity "${entry.tableName}" {\n    ${columns.join(
          "\n    --\n    "
        )}\n}`;
      })
      .join("\n");
  }
}
