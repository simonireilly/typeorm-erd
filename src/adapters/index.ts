import { DataSource } from "typeorm";
import { MermaidErd } from "./mermaid";
import { PlantUMLErd } from "./plantuml";

const BUILDERS = {
  mermaid: MermaidErd,
  plantuml: PlantUMLErd,
};

type TBuildTypes = keyof typeof BUILDERS;

class ERDBuilder {
  private instance: MermaidErd | PlantUMLErd;

  constructor(public instanceType: TBuildTypes, dataSource: DataSource) {
    const adapter = BUILDERS[instanceType];
    this.instance = new adapter(dataSource);
  }

  public async initialize() {
    return this.instance.initialize();
  }

  public render(): string {
    return this.instance.render();
  }
}

export { PlantUMLErd, MermaidErd, ERDBuilder };
