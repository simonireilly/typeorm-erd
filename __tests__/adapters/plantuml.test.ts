import { AppDataSource } from "../../examples/basic/src/data-source";
import { AppDataSource as JoinDataSource } from "../../examples/join-tables/src/data-source";
import { AppDataSource as CartData } from "../../examples/shopping-cart/src/data-source";
import { PlantUMLErd } from "../../src/adapters";

describe("mermaid adapter", () => {
  describe("integration tests", () => {
    test("renders the complete basic tables", async () => {
      const erd = new PlantUMLErd(AppDataSource);
      await erd.initialize();

      expect(erd.render()).toMatchSnapshot();
    });

    test("renders the complete cart tables", async () => {
      const erd = new PlantUMLErd(CartData);
      await erd.initialize();

      expect(erd.render()).toMatchSnapshot();
    });
  });

  describe("public methods", () => {
    it("renderRelations", async () => {
      const erd = new PlantUMLErd(AppDataSource);

      await erd.initialize();

      expect(erd.buildRelations()).toEqual(["comment }|--|| user"]);
    });

    it("renderRelations", async () => {
      const erd = new PlantUMLErd(JoinDataSource);

      await erd.initialize();

      expect(erd.buildRelations()).toEqual([
        "tag ||--|{ product_tag",
        "product ||--|{ product_tag",
      ]);
    });
  });
});
