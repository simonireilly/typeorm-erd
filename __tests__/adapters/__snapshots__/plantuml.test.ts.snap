// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mermaid adapter integration tests renders the complete basic tables 1`] = `
"@startuml
skinparam linetype ortho
entity \\"user\\" {
    id : INTEGER : <<PK>> : //\\"The primary user identifier\\"//
    --
    firstName : VARCHAR
    --
    lastName : VARCHAR
    --
    age : INTEGER
}
entity \\"comment\\" {
    id : INTEGER : <<PK>> : //\\"The ID for the comment\\"//
    --
    text : VARCHAR(300)
    --
    likes : INTEGER
    --
    authorId : INTEGER : <<FK>> : //\\"The primary user identifier\\"//
}
comment }|--|| user
@enduml"
`;

exports[`mermaid adapter integration tests renders the complete cart tables 1`] = `
"@startuml
skinparam linetype ortho
entity \\"cart_item\\" {
    id : BIGINT : <<PK>>
    --
    productId : BIGINT : <<FK>>
    --
    cartId : BIGINT : <<FK>>
    --
    sku : VARCHAR(100)
    --
    price : FLOAT
    --
    discount : FLOAT
    --
    quantity : SMALLINT
    --
    active : TINYINT
    --
    createdAt : DATETIME
    --
    updatedAt : DATETIME
    --
    content : TEXT
}
entity \\"category\\" {
    id : BIGINT : <<PK>>
    --
    parentId : BIGINT : <<FK>>
    --
    title : VARCHAR(75)
    --
    metaTitle : VARCHAR(100)
    --
    slug : VARCHAR(100)
    --
    content : TEXT
}
entity \\"product_meta\\" {
    id : BIGINT : <<PK>>
    --
    productId : BIGINT : <<FK>>
    --
    key : VARCHAR(50)
    --
    content : TEXT
}
entity \\"product_review\\" {
    id : BIGINT : <<PK>>
    --
    productId : BIGINT : <<FK>>
    --
    parentId : BIGINT : <<FK>>
    --
    title : VARCHAR(100)
    --
    rating : SMALLINT
    --
    published : TINYINT
    --
    createdAt : DATETIME
    --
    publishedAt : DATETIME
    --
    content : TEXT
}
entity \\"tag\\" {
    id : BIGINT : <<PK>>
    --
    title : VARCHAR(75)
    --
    metaTitle : VARCHAR(100)
    --
    slug : VARCHAR(100)
    --
    content : TEXT
}
entity \\"product\\" {
    id : BIGINT : <<PK>>
    --
    userId : BIGINT : <<FK>>
    --
    title : VARCHAR(75)
    --
    metaTitle : VARCHAR(100)
    --
    slug : VARCHAR(100)
    --
    summary : TINYTEXT
    --
    type : SMALLINT
    --
    sku : VARCHAR(100)
    --
    price : FLOAT
    --
    discount : FLOAT
    --
    quantity : SMALLINT
    --
    shop : TINYINT
    --
    createdAt : DATETIME
    --
    updatedAt : DATETIME
    --
    publishedAt : DATETIME
    --
    startsAt : DATETIME
    --
    endsAt : DATETIME
    --
    content : TEXT
}
entity \\"order_item\\" {
    id : BIGINT : <<PK>>
    --
    productId : BIGINT : <<FK>>
    --
    orderId : BIGINT : <<FK>>
    --
    sku : VARCHAR(100)
    --
    price : FLOAT
    --
    discount : FLOAT
    --
    quantity : SMALLINT
    --
    createdAt : DATETIME
    --
    updatedAt : DATETIME
    --
    content : TEXT
}
entity \\"transaction\\" {
    id : BIGINT : <<PK>>
    --
    userId : BIGINT : <<FK>>
    --
    orderId : BIGINT : <<FK>>
    --
    code : VARCHAR(100)
    --
    type : SMALLINT
    --
    mode : SMALLINT
    --
    status : SMALLINT
    --
    createdAt : DATETIME
    --
    updatedAt : DATETIME
    --
    content : TEXT
}
entity \\"order\\" {
    id : BIGINT : <<PK>>
    --
    userId : BIGINT : <<FK>>
    --
    sessionId : VARCHAR(100)
    --
    token : VARCHAR(100)
    --
    status : SMALLINT
    --
    subTotal : FLOAT
    --
    itemDiscount : FLOAT
    --
    tax : FLOAT
    --
    shipping : FLOAT
    --
    total : FLOAT
    --
    promo : VARCHAR(50)
    --
    discount : FLOAT
    --
    grandTotal : FLOAT
    --
    firstName : VARCHAR(50)
    --
    middleName : VARCHAR(50)
    --
    lastName : VARCHAR(50)
    --
    mobile : VARCHAR(15)
    --
    email : VARCHAR(50)
    --
    line1 : VARCHAR(50)
    --
    line2 : VARCHAR(50)
    --
    city : VARCHAR(50)
    --
    province : VARCHAR(50)
    --
    country : VARCHAR(50)
    --
    createdAt : DATETIME
    --
    updatedAt : DATETIME
    --
    content : TEXT
}
entity \\"user\\" {
    id : BIGINT : <<PK>>
    --
    role : ENUM
    --
    firstName : VARCHAR(50)
    --
    middleName : VARCHAR(50)
    --
    lastName : VARCHAR(50)
    --
    mobile : VARCHAR(15)
    --
    email : VARCHAR(50)
    --
    passwordHash : VARCHAR(32)
    --
    admin : TINYINT
    --
    vendor : TINYINT
    --
    registeredAt : DATETIME
    --
    lastLogin : DATETIME
    --
    intro : TINYTEXT
    --
    profile : TEXT
}
entity \\"cart\\" {
    id : BIGINT : <<PK>>
    --
    userId : BIGINT : <<FK>>
    --
    sessionId : VARCHAR(100)
    --
    token : VARCHAR(100)
    --
    status : SMALLINT
    --
    firstName : VARCHAR(50)
    --
    middleName : VARCHAR(50)
    --
    lastName : VARCHAR(50)
    --
    mobile : VARCHAR(15)
    --
    email : VARCHAR(50)
    --
    line1 : VARCHAR(50)
    --
    line2 : VARCHAR(50)
    --
    city : VARCHAR(50)
    --
    province : VARCHAR(50)
    --
    country : VARCHAR(50)
    --
    createdAt : DATETIME
    --
    updatedAt : DATETIME
    --
    content : TEXT
}
entity \\"product_category\\" {
    categoryId : BIGINT : <<PK>>
    --
    productId : BIGINT : <<PK>>
}
entity \\"product_tag\\" {
    productId : BIGINT : <<PK>>
    --
    tagId : BIGINT : <<PK>>
}
cart_item }|--|| cart
cart_item }|--|| product
category }|--|| category
category ||--|{ product_category
product_meta }|--|| product
product_review }|--|| product_review
product_review }|--|| product
tag ||--|{ product_tag
product }|--|| user
product ||--|{ product_category
product ||--|{ product_tag
order_item }|--|| order
order_item }|--|| product
transaction }|--|| order
transaction }|--|| user
order }|--|| user
cart }|--|| user
@enduml"
`;
